### PR TITLE
Documentation for PerTickCache

### PR DESCRIPTION
**Pull request by antonykiran10**

### CODE REVIEW

File: source/Optimization.rst [hunk 1]
```diff
@@ -0,0 +1,37 @@
+Optimisation
+===============
+
+
+PerTickCache
```
Comment (dpcherian) [inline source/Optimization.rst:5]: First please describe what the per-tick cache actually _is_. Go through the ``documentation.md`` file, in the "Help and FAQs" section there's a brief description of the per-tick cache that you can start off by describing. Secondly, why are you capitalising it like this? You aren't really describing a computational _function_ here, more a feature.

File: source/Optimization.rst [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+----------------
+This function is used to retrieve values from the cache if some value is already computed. If the value is not present in the cache, it computes the value using a user-defined function and stores it in the cache.
```
Comment (dpcherian) [inline source/Optimization.rst:7]: You have forgotten to mention the most important thing! The per-tick cache is only valid for **one tick**!

File: source/Optimization.rst [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+This is extremely useful in reducing the simulation time when we repeatedly have to query the context to compute the value for every agent in a tick.
+
+For example, consider the case in the simple SIR model where we have to find the fraction of infected individuals around an agent in a particular place. Here the infected fraction of all agents in the same place would be the same. So while running the simulation, we have to compute the same value repeatedly for all agents in that place. We need to query the context for all agents during each tick, thereby increasing the simulation time and computational resource required.
+Here when we implement PerTickCache, the infected fraction will be computed for all places as and when needed during each tick, and that value will be stored in the cache to be used for all the agents in that place.
```
Comment (dpcherian) [inline source/Optimization.rst:11]: Please describe in some detail how the value is stored. For example, the modeller has to define a "unique key", otherwise the the value that is stored will not be accurately stored. This is explained in the ``documentation.md`` file, but you will need to reword it to make it a little more accessible. 

Please give an example of this as well. For example, in the example you give below, suppose you used the ``uniqueKey`` as just being ``placeType``, what would happen?

File: source/Optimization.rst (hunk 2 above)
Comment (dpcherian) [inline source/Optimization.rst:7]: It isn't really a function, it's more a class in our example. But for now, you can just call it a "feature". Please go through the ``documentation.md`` file in the BharatSim repo that you have.

Comment (dpcherian) [reply source/Optimization.rst:7]: You have mentioned it briefly later, but you should really stress this here as well.

File: source/Optimization.rst [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+If the value is not computed for a place, the PerTickCache function will compute the value using a user-defined function. 
+
+Example for PerTickCache implementation:
```
Comment (dpcherian) [inline source/Optimization.rst:15]: Very good idea. However, you are writing in markdown! You don't need to put in a chunk of code with comments inside it. Please reformat this so that you first describe what a function is doing in markdown, and then just include the code for that function below it. Do these for both the comments below: basically, be a little more verbose. After you are done, explain again that the ``computeTheValue`` function is only called **once** per tick per location, and how that can speed things up when you have a lot of agents in a location.

**Review by dpcherian — CHANGES_REQUESTED** [review]
Very good, some small comments that I would like you to quickly fix before I merge them. Please make sure you go through the ``documentation.md`` file in the BharatSim repository that you have with you, it'll help you improve this.

File: source/Framework Basics.rst [hunk 5]
```diff
@@ -0,0 +1,173 @@
+Framework Basics
+=================
+
+Interventions
+-------------
+Interventions can help you execute specific actions based on the current state of the simulation. The simulation engine of Bharat-Sim checks if an inactive intervention can be activated based on the given conditions during every tick. All the active interventions will be recorded in the context and the user can query the context using ``context.activeInterventionNames.contains(interventionName)``.
```
Comment (dpcherian) [inline source/Framework Basics.rst:6]: Please use the correct capitalisation of BharatSim (as opposed to Bharat-Sim).

File: source/Framework Basics.rst [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+There are 4 different kinds of Interventions available in Bharat-Sim:
+
+1.Intervention
+~~~~~~~~~~~~~~~~~~~~~
+This is a Generic intervention that can be invoked several times or once when the activation condition is met based on the parameters passed.
+This can be used to implement lockdows during an epidemic everytime the infected fraction of agent reaches a certain threshold.
+There are 5 parameters that can be passed ``Intervention`` object:
+
+    * ``interventionName``: This is the unique name of the intervention.
```
Comment (dpcherian) [inline source/Framework Basics.rst:16]: Please specify which parameters are compulsory, and which are optional. Do this for the rest of the document as well.

File: source/Framework Basics.rst [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    * ``shouldActivateFunc``: This function which tells whether this intervention should be activated or not. The function should return a Boolean value.
+    * ``shouldDeactivateFunc``: This function which tells whether this intervention should be deactivated. The function should return a Boolean value.
+    * ``firstTimeActionFunc``: This is an optional function which will be executed at the start of the intervention
+    * ``whenActiveActionFunc``: This is an optional function which will be executed per tick when intervention is active.
+
+Example using the ``Intervention`` object:
```
Comment (dpcherian) [inline source/Framework Basics.rst:22]: Same comment as I made earlier: explain the parts of this code block in markdown, don't just use code comments. Furthermore, please explain where this ``addLockdown`` function is actually called, and why we are placing all of this inside a function.

File: source/Framework Basics.rst [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+````````````````````````````````````````````````````
+.. code-block:: scala
+
+    //This implements a lockdown when the nuber of agents in the infected compartment is greater than or equal to 2000
+    private def addLockdown(implicit context: Context): Unit = {
+
+    var interventionActivatedAt = 0
+    val interventionName = "lockdown"
+    val activationCondition = (context: Context) => getInfectedCount(context) >= 2000
```
Comment (dpcherian) [inline source/Framework Basics.rst:31]: What is the `getInfectedCount` function? A beginner might not understand that this is not a framework defined function. Explain that this is user-defined, and you can link to the "Writing your first program" section and say that it will be explained in greater detail there.

File: source/Framework Basics.rst [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    val firstTimeExecution = (context: Context) => interventionActivatedAt = context.getCurrentStep
+    val deActivationCondition = (context: Context) => {
+    context.getCurrentStep >= interventionActivatedAt + 2*14
+    }
+    val intervention =
+    Intervention(interventionName, activationCondition, deActivationCondition, firstTimeExecution)
```
Comment (dpcherian) [inline source/Framework Basics.rst:37]: Lines 36, 37, and 41 are the most important in this code block. Keep them together, and move the ``lockdownSchedule`` to just before the ``registerSchedules`` command.

File: source/Framework Basics.rst [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+    //This implements a lockdown when the nuber of agents in the infected compartment is greater than or equal to 2000
+    private def addLockdown(implicit context: Context): Unit = {
+
+    var interventionActivatedAt = 0
+    val interventionName = "lockdown"
+    val activationCondition = (context: Context) => getInfectedCount(context) >= 2000
+    val firstTimeExecution = (context: Context) => interventionActivatedAt = context.getCurrentStep
+    val deActivationCondition = (context: Context) => {
+    context.getCurrentStep >= interventionActivatedAt + 2*14
```
Comment (dpcherian) [inline source/Framework Basics.rst:34]: Why is there a ``2*14`` here?

File: source/Framework Basics.rst [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+This can be used 
+
+There are 5 parameters that can be passed to the ``OffsetBasedIntervention`` object:
+
+    * ``interventionName``: This is the unique name of the intervention.
+    * ``shouldActivateWhen``: This function decides when should the intervention be activated.
+    * ``endAfterNTicks``: This is the offset 'n'; after n ticks from the start tick simulation will end.
+    * ``firstTimeActionFunc``:This is an optional function which gets executed when simulation starts.
+    * ``whenActiveActionFunc``: This is anoptional function which gets executed per tick when simulation is active.
+
```
Comment (dpcherian) [inline source/Framework Basics.rst:112]: No example for `OffsetBasedIntervention`?

File: source/Framework Basics.rst [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+There are 5 paramertes that can be passed to the ``IntervalBasedIntervention`` object:
+
+    * ``interventionName``: This is the unique intervention name.
+    * ``startTick``: This ia an integer that specifies the start tick for intervention (inclusive); it should not be greater than endTick.
+    * ``endTick``: This is an integer that specifies the end tick for the intervention (exclusive, intervention will not be active at "endTick".)
+    * ``firstTimeActionFunc``: This is an optional function which gets executed when simulation starts.
+    * ``whenActiveActionFunc``: This is an optional function which gets executed per tick when simulation is active.
+
+Example suing the ``IntervalBasedIntervention`` object:
```
Comment (dpcherian) [inline source/Framework Basics.rst:68]: Be careful with your spelling, you've made this mistake everywhere below as well.

**Review by dpcherian — CHANGES_REQUESTED** [review]
Very good, but some important changes to be made:

1. As I mentioned before, please don't just paste in a chunk of code with comments, it's not going to help anyone. Try to write a little bit in markdown describing what you're doing every few lines. You can then add the "complete" code at the end of the sub-section.
2. Some of the examples are **far** too complex! Please simplify them. Make them "toy" examples that people can understand. Then, when they look at the examples in the simulation engine, they will be able to understand them better.  For example, even I have no real idea what's going on in the ``SingleInvocationIntervention``, and a beginner will be far too intimidated by all of this. 

I would suggest using only **one** type of intervention (say, a lockdown) and showing how it can be implemented in multiple ways (as a bare ``Intervention``, as an ``IntervalBasedIntervention``, as an ``OffsetBasedIntervention``, and as a ``SingleInvocationIntervention``. That way, a beginner will also have a better sense of when to use which type of intervention.

File: source/Framework Basics.rst (hunk 5 above)
Comment (antonykiran10) [reply source/Framework Basics.rst:6]: Actually, this section is still in progress. I thought that git push means it will push only to my forked repo. I don't remember creating a pull request after I modified the Intervention.rst file. I  will look into all the suggestions and will definitely do a spell check before I create a pull request again.

Comment (dpcherian) [reply source/Framework Basics.rst:6]: No problems, just make these small changes and I'll merge them. Things are mostly ok!

**Comment by dpcherian** [comment]
Hey @antonykiran10 could you please make these modifications quickly? Don't worry if they're not perfect, just finish as much a possible, please. I'd really like to merge it and start adding to it.

**Comment by dpcherian** [comment]
@antonykiran10 When you're done with the edits, please "request a review" from me, or tag me in a comment so that I can approve the changes.

**Review by dpcherian — APPROVED** [review]
Great, thanks!
